### DISCUSSION TITLE
What's the future of Actual? A roadmap

### DISCUSSION THREAD

**Discussion by jlongster** [actualbudget/releases#112] [Announcements]
With the recent interest in Actual I wanted to write down both a short-term and long-term roadmap. I've been thinking about doing this for a long time, and it's good to get all this down clearly in one single place so that we're all on the same page. I haven't been as active in the last couple of days while focusing on this, but as you can see below, there is a **lot** to come!

# First, some context

To understand the future, first we need to understand the past.

A quick bit of history: I started Actual years ago. I just checked the git history and found the first commit:

```
commit 5a2e204a465ce25ade910d30b7991c130e3e37cd
Author: James Long <redacted>
Date:   Thu Mar 10 17:48:58 2016 -0500

    initial
```

March, 10, 2016! That's almost 6 years! A lot happened that I won't go into now, but roughly the first couple years was just hacking on ideas for fun. It wasn't until 2018 where I got more serious about it and started working through what it means to be a [local-first app](https://www.inkandswitch.com/local-first/).

This is when I developed a syncing engine and a mobile app to go with it. I saw a _lot_ of potential here and decided it was worth investing in. After a lot of work, I launched on [Product Hunt](https://www.producthunt.com/posts/actual-budget) in January of 2019 and got #4 for the day. In 2019 I worked full-time on it and over the last year and half part-time, adding a lot more features.

That's where we are today. While I'm proud of the app and so happy that people are noticing how thoughtful it is, in my eyes it's still a glimpse of what it could be. A huge amount of the work so far has been foundational: setting up a local-first architecture that will unlock things nobody else can do.

## Investment in a local-first app

Being a local-first app requires a _huge_ investment. The syncing engine itself was not too difficult — the difficulty is in all the little things you don't think about. The entire UX flow for creating/managing files is different. I can't easily use 3rd party libraries off-the-shelf because they just assume a server (and sharing all your data). The implications are huge, and it's totally worth it, but there's no doubt it's a huge investment.

The good thing is that I believe most of the investment in a local-first architecture is done. From here on it's just maintenance like any other feature. However, there are still constraints that slow down development, the biggest one being that I can't easily change the database schema. Normal webapps can easily deploy a new version and change the backend database at the same time. When I want to change something, I have to maintain backwards compatibility and deploy the change super carefully. This probably makes it take 1.5-2x as long to develop new fundamental features.

I bring that up so you can understand why certain features take a while. I can't quickly go in there and just… change stuff. I need to think through the implications!

But remember what you get for this. **There are two things that local-first apps make possible:**

* **Data ownership and privacy**. You own your data; you can locally run queries against it or whatever you want. While we offer a syncing service, you can encrypt your data end-to-end for complete privacy.
* **It's fast**. I don't care how much you optimize your server, if you have to make network requests _it's going to be slow_. All it takes is a slow connection to the wifi router in your house; you can't solve that! When all of your data is local by default, every single operation in the app is consistently surprisingly fast.

## Multiple platform support

For better or for worse, Actual is available on every single major platform: the web, desktop apps for macOS, Windows, and Linux, and mobile apps for iOS and Android. I say "for worse" because building out these apps absolutely slowed down development as well.

Actual started out as just a desktop app. It _had_ to be a desktop app because it used a local sqlite file, something you can't do on the web. As I developed the syncing engine, it felt natural to make a mobile app to go with it.

The web version is pretty new. I launched an experimental version some time ago, but it wasn't until this August that I developed [a new technology](https://jlongster.com/future-sql-web) that makes it viable. On the web the entire app is still local, and my work made it possible to continue using a sqlite file like all other platforms.

Because of this history, things like onboarding aren't as smooth as they should be. I'm in the process of reworking the website to take the web version better into account, which solves some of the awkward onboarding.

The release process also needs to be improved. The web version allows me to immediately deploy fixes which is extremely powerful compared to the desktop apps which require a bulky release process. In light of this, I want to move to a rolling release cycle where the web version is continuously deployed, and the desktop/mobile apps automatically get released on a certain cycle that includes all of the latest fixes.

The "ship it fast" and "get a minimal app out" people would consider this process a mistake. I should have developed and gotten more users on a single platform first before supporting others. This is short-sighted thinking. You might be able to dig a hole fast, but then you're stuck in a hole. Some things require more deliberate, long-term thinking before you start digging. It would have been incredibly hard to transition to multi-platform support later, and the project would likely have failed. Now that the architecture is proven, we can focus on building features.

## A new way to build the frontend

The largest amount of technical debt currently exists in the frontend. While the app works well, working on the frontend requires careful wiring up. It only works well because I'm detail-oriented and rigorously manually test every single interaction when I'm working on it.

The biggest thing holding the app back is not being able to quickly build UIs that come together nicely by default. This means everything from design, accessibility, motion, information architecture, and more.

Another major pain point: the mobile app is an entirely separate UI. I took this approach because I believed the mobile experience to be so different. However, this ends up duplicating so much design work and I'm come to appreciate thinking about the UI as a single responsive interface.

A concrete example: I plan on replacing the current table component entirely. The new one will be more accessible, allow far more customization (show/hide columns, etc), and work on both desktop and mobile.

I'm going to work on a new way to build the frontend that is more composable and provides a great UI without much work. In all this work **I'll be leveraging 3rd party libraries** which will greatly increase the development speed. Ideally much of this would work the same on desktop and mobile, so the mobile app will have access to more features. I'll be able to build a feature once, and it will work great out of the box on both platforms. **This is not a rewrite**, but will be incrementally adopted.

Overall, there's been a _ton_ of investment in just making the app work in this unique local-first architecture. I believe this will pay off, and there's some work to do to bring everything together in a consistent way.

# The Roadmap

Now let's look to the future. Given how much time a local-first approach has required, and how many features have been added over time, the app is not as consistent and intuitive as I'd like. Naturally there are also missing features.

The future is about adding new features, tying all these features together in a consistent way, and in the long-term working towards an even better way to explore your finances.

## Short-term

**Timeframe**: from now to end of 2021

* Ever since I opened up the [issue tracker](https://github.com/actualbudget/releases/issues), you all have been filing great bugs. Thank you! I haven't replied to many of them, but know that I see it and it's _incredibly_ useful to have these bug reports. The goal is simply to **fix as many reported bugs as possible**. I can't guarantee which ones will be fixed but a good number of them should be.

* **Make YNAB4/5 importing solid.**

	Integrate the import-ynab5 project directly in the app and address any issues. YNAB4 importing _should_ already be good. I'll fix the problem where you don't see budget values right after importing (requires an app restart). In fact this is already fixed, and I'll do a release soon.

* **Improve transactions generally.** There are several small features that are long overdue:
	* Ability to add payees to split transactions
	* Improving the UX of transfers
	* Adding a new category directly from the dropdown
	* Add more CSV import options
	* Many other small things (like showing the running balance for scheduled transactions)
	* Scheduling splits and transfers

* **Improvements to local files.**
	
	One long-standing problem on macOS is the default location to store files is `~/Documents`. If you have iCloud syncing turned on, it occasionally messes with your data. The default location should change so users don't ever experience this problem. Also, there are some big performance fixes possible for the web storage layer.

* **Allow annual plans.**
	
	An often requested feature. I'll add the ability to pay annually at the same price.

## Long-term

I'm not going to give definitive dates for any of this yet. We'll see where we are at the end of 2021 and I'll post another roadmap specifically for Q1 2022 when that comes around. Features marked with `*` are likely to happen sooner rather than later, but things can easily change.

There are definitely things missing from here. If you don't see something here, it doesn't mean it won't happen. These are just the things I'm thinking about.
  
* **Adding reconciled as a transaction status** *
  
  This has been requested many times. It would mark transactions as reconciled and discourage editing them, and improve the reconciliation workflow.

* **Better credit card support** *

  This is also a common piece of feedback. There are some small things here, like improving the "create account" workflow that makes it clearer how credit card balances are handled. Right now you have to know to put in a negative amount for the starting balance for it to work correctly.
  
  It also covers improving the overall credit card workflow. Not sure what this looks like yet; I don't want to complicate how the budget numbers are calculated. It's already difficult enough to track down problems if something is off. But it's clear we need to improve how to track a long-running credit card debt that needs to be paid down.
  
* **More report options** *

	There is so much potential in reports because all your data is available locally. Custom reports is a different feature listed below; this is more about improving the builtin reports until custom reports are available.
	
	The existing reports are going to be completely rebuilt from scratch, with the ability to apply any filters on the included transactions. Ideally they would render on mobile as well (see UI revamp below).
  
* **Goals** *

	Another often-requested feature! I don't have any clear plans for this yet; there has been a lot of great feedback and I need to go through it all to understand the use cases. From there we'll find an appropriate implementation.
	
* **UI revamp** *

	This will be a long-running effort to overhaul the UI generally. This includes updated colors, improving accessibility, replacing core components, and consolidating the desktop and mobile experience.
	
	One specific thing to call out here: This will include a new table that allows **customizing and resizing columns** and a better searching/filtering experience. This impacts almost every screen in the app, but this will particularly revamp the transaction editing experience.
	
	A big goal is to make it possible to share more code across desktop and mobile. This will drastically the reduce the amount of effort needed to support features on both.
	
* **Dark mode and themes**
  
  Related to the UI revamp, the new system and updated colors will support a dark theme and potentially custom themes.

* **Mobile app improvements**

	Very related to the UI revamp, but it's worth calling out that with all the other work going on, it will make it much easier to bring the mobile app to be on par with the desktop app feature-wise. In particular, the budget screen will be getting a lot more features like showing transactions specific to a category for a month.

* **Transaction tags**

  In the `notes` field, you should be able to type something like `#gift` to tag the transaction with `gift`. Tags would drastically increase the flexibility of the system, allowing you to create custom workflows for tracking anything you want.

* **Bank syncing**

	This allows you to connect with your bank and automatically download transactions. There's no question that this will be a huge improvement, and this could be a turning point for the app to gain a lot more users. It's a little ways off because I want to get the app to a more stabilized and consistent first.
	
	It's worth noting that **all the current UX decisions take bank syncing in mind**. 

* **Multi-currency support**

	Support different currencies for each account, and automatically convert between them when querying across accounts. I'm not aware of any other app that does zero-sum budgeting with multi-currency support. It's complicated, but I have some ideas for how to make it work.

* **Localized UI**

	Display the app in various languages.

* **Custom reports**

	This is a big one. I've had this in mind since the app's conception, and it's a driving reason for the local-first architecture.
	
	Imagine being able to query your data and visualize it any way you want. I'm still working through what how this works in reality, but I've already built [ActualQL](https://actualbudget.com/docs/developers/ActualQL/) which is a simple query language to asking questions about your data. With custom reports, you'd be able to take such queries and run them right in the app, visualizing them in interesting ways.
	
	This is a power user feature, but the really cool part is that custom reports could easily be exported and shared. We'd be able to grow a large library of reports built entirely by the community, and users who don't know how to write queries would still benefit from all the available reports.

* **Off-budget categories**

	There are several use cases when you just want to track transaction off the budget, but it's a shame that doing so you lose out on categories. You can't categorize off-budget transactions, which means you can't track them in detail and see them in various reports.
	
	I've been thinking about the idea of off-budget categories instead of off-budget accounts, and maybe on-budget accounts simply make it hard to use off-budget categories (but you still could). This _could_ be related to credit cards; you could categorize the **debt** part of a credit as off-budget and track it somewhere else, while spending uses on-budget categories. Not sure yet, but I think it's interesting.
	
	We have to be careful about this; we don't want to make it easy to spend money off the budget. However, relaxing the rules a little allows you to flex your workflows in ways that works for you. Ultimately, we trust _you_ to make the right decisions and we aren't going to dictate your life.
 

* **Multiple budgeting types**

	Off-budget categories is a good segue into this: I've been thinking about making the app more useful outside of the zero-sum budgeting context. It just doesn't make sense to have an incredibly powerful app with bank syncing and custom reports (eventually), only to be useful for people who use zero-sum budgeting.
	
	The budget itself it just internally a bunch of queries. It's relatively easy to write a different budgeting algorithm, and I've been thinking about providing a simple income vs expense budget. This type of budget would not force you to reconcile spending; it would show you how much you overspent and that's it. We could probably show some statistics how often  you're overspending and sum up how much you've overspent year-to-date and things like that.
	
	Before you freak out: zero-sum budgeting will always be a core feature of the app. It'd be ridiculous for me to pivot away from that when almost all users are currently here _for_ it. Everything in the app will always take it into account.
	
	Look at it this way: a simpler budget could be a way to onboard people onto the app who _don't_ use zero-sum budgeting, and then we can teach them about it. It could be a good on ramp to sell that budgeting method.

* **Fluid UI**

	This one is more conceptual. It will depend on the UI revamp and other features, and this is probably a long ways off, but I wanted to pitch some ideas.
	
	I love apps where the UI is "fluid". Another term I've been using is "transparent UIs". These are UIs that lack a rigid structure and let the user radically change them.
	
	Two examples of this are spreadsheets and [Roam](https://roamresearch.com)-style apps. Open a spreadsheet and for the most part you just see a grid of cells. Open Roam and you just see a notepad. It looks simple at first glance, but anyone who's used either knows that underneat the simple UI is an extremely powerful engine capable of all kinds of things. This engine is exposed via a moldable UI that lets you "program" in a structured way (a 2d grid of cells for spreadsheets, an outline of taggable text in Roam).
	
	Fun fact: in the early days (~2016) I started building Actual in this way. It was a set of tables that took queries and allowed arbitrary editing. It was very spreadsheet-inspired, but I gave up once I realized I was building a spreadsheet (extremely complicated, and might as well use Excel).
	
	However, there are still UI design lessons we can learn from those style of apps. I think this is reflected in the current design of Actual; it's frequent to here from users feedback like "wow I thought the app was too simple, but now that I started using I see how deep it is". This is the best kind of feedback I can get; it's exactly what I was going for!
	
	Obviously there's a balance to be made here: looking a grid of cells is daunting. You don't want to force the user to build everything themselves (which is why we all don't use a spreadsheet). You also want things to be discoverable; seeing a list of actions make it obvious what can be done. However, you can't just list _all_ the actions on the screen because it would overwhelm the UI.
	
	I'm inspired by these kinds of UI, and I want to find a balance where Actual feels programmable but also discoverable and intuitive.
	
	**Custom reports** is a key part of this. You'll be able to build a lot of features yourself! I'm not sure how far we can go with it, but it would be awesome if it allows the app itself to become programmable.
	
	Here's some mockups I made to demonstrate what a more fluid UI would look like. Right now Actual's UI is rigid; you can't easily pull up part of the app without navigating directly to it. I frequently find myself deep in the transaction screen, and then I want to find look at some other data from the budget or reports.
	
	What if you could pull up any page as a right-hand sidebar? Here I pull up the reports to the right, or maybe I want to look at the budget. Putting this in context of the transactions is extremely useful:
	
	![Kapture 2021-11-24 at 14 18 00](https://user-images.githubusercontent.com/17031/143304249-a9b7376e-6a8e-4ec6-a373-72c0352b7006.gif)

	What if the data needs more space though? If I wanted to open the transactions table without losing my place, the table wouldn't fit on the right. What if shift+click opened any page _on top_ of the current page instead of navigating to it? Here I open up another account to look for a transaction on top of an existing account:

	![Kapture 2021-11-24 at 14 18 52](https://user-images.githubusercontent.com/17031/143304308-48e85a03-ad25-416b-b9c7-2f552c174456.gif)
	
	A great use case for this would be clicking the "spent" amount in the budget table to see the transactions that make it up. Right now it takes you away from the budget, but you'd be able to open up transactions on top instead.
	
	A third idea: what if the sidebar was more dynamic? Currently it always lists the accounts in a specific order. What if you could widen it and render any data in there, maybe from a custom report?

	![Kapture 2021-11-24 at 14 19 38](https://user-images.githubusercontent.com/17031/143304334-f125a43e-a7b5-4552-8b6d-b7b9ce723baf.gif)

	I'd love to break down the UI into more composable workflows like this.
	
	I'm interested to hear any comments about this! Any big features that I missed?

**Comment by pepesenaris** [comment]
Docs… a bit more content on the documentation, even if it’s a few links around how to do zero sum budgeting and then more examples on how to use the app features to implement cases like credit cards, off budget retirement accounts, regular savings,etc. 

It could serve both to newcomers and “veterans” that may get to play with new workflows for their use cases

**Reply by nikohd12** [reply]
> Docs… a bit more content on the documentation, even if it’s a few links around how to do zero sum budgeting and then more examples on how to use the app features to implement cases like credit cards, off budget retirement accounts, regular savings,etc.
> 
> It could serve both to newcomers and “veterans” that may get to play with new workflows for their use cases

I would agree but also, given that he's a one-man team, it would be best to just rely on the community knowledge first. I'm not sure if this is allowed or whatever, but YNAB provides good YouTube videos about zero-sum budgeting.

For the app specifically, we might need community volunteers to document.

**Reply by bpaulien** [reply]
There's a wiki tab right here on the github site. Would that be a good place to start a community lead documentation effort?

**Reply by pepesenaris** [reply]
All 🙌 for a community based docs. It could be as simple as adding a link to official docs, as that could be where new people start with, to Github discussions or any other platform

**Reply by jlongster** [reply]
Docs are a great point. I'm totally down for getting help from the community, but I think we'll need some guardrails for consistency. Or at least, a way to turn docs from the community into official docs. Let me think about this a little bit! I bet we can set something up.

**Reply by MiriamNZ** [reply]
Getting the community to write more support docs sounds a good idea to me. The more we can do to help the better. But you are right, that you would still need to check/correct/approve.

**Comment by MiriamNZ** [comment]
The more fluid UI looks really useful. Seeing how a transaction affects the budget categories would be really helpful.

**Comment by alehel** [comment]
Really curious why this isn't also a blog post on the actual budgets website. The fact that there have been no posts since august gives people the impression that there's nothing going on. At the very least, release notes should be published there.

**Reply by Pikafu** [reply]
I second this notion. I thought nothing was happening with Actual until I stumbled across this post in the Discussions section. It would be nice to at least link this post on the Actual website.

**Reply by Kidglove57** [reply]
Just to chip in and say that you can  now see the release notes on the Blog section on the website! Just click on the relevant release number in the Blog to see more details. Here is the link:

https://actualbudget.com/blog/
